src/borrow_tracker/tree_borrows/tree/tests.rs

```diff
@@ -0,0 +1,674 @@
+            let x = self.x.perform_access(acc.kind, xrel)?;
+            let y = self.y.perform_access(acc.kind, yrel)?;
+            Ok(Self { x, y })
+        }
+
+        fn end_protector_x(&self, _xy_rel: RelPosXY) -> Result<Self, ()> {
+            let x = self.x.end_protector();
+            let y = self.y.clone();
+            Ok(Self { x, y })
+        }
+
+        fn end_protector_y(&self, _xy_rel: RelPosXY) -> Result<Self, ()> {
+            let x = self.x.clone();
+            let y = self.y.end_protector();
+            Ok(Self { x, y })
+        }
+
+        fn perform_test_event<RetX, RetY>(
+            &self,
+            evt: &TestEvent<RetX, RetY>,
+            xy_rel: RelPosXY,
+        ) -> Result<Self, ()> {
+            match evt {
+                TestEvent::Access(acc) => self.perform_test_access(acc, xy_rel),
+                // The fields of the `Ret` variants just serve to make them
+                // impossible to instanciate via the `!` type; we can
+                // always ignore their value.
+                TestEvent::RetX(_) => self.end_protector_x(xy_rel),
+                TestEvent::RetY(_) => self.end_protector_y(xy_rel),
+            }
+        }
+    }
+
+    impl Exhaustive for LocStateProtPair {
+        fn exhaustive() -> Box<dyn Iterator<Item = Self>> {
+            Box::new(<[LocStateProt; 2]>::exhaustive().map(|[x, y]| Self { x, y }))
+        }
+    }
+
+    impl fmt::Display for LocStateProtPair {
+        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+            write!(f, "x:{}, y:{}", self.x, self.y)
+        }
+    }
+
+    /// Arbitrary sequence of accesses, as experienced by two mutually foreign pointers
+    /// to the same location.
+    #[derive(Debug, Clone)]
+    struct OpaqueCode<RetX, RetY> {
+        events: Vec<TestEvent<RetX, RetY>>,
+    }
+
+    impl<RetX, RetY> fmt::Display for OpaqueCode<RetX, RetY> {
+        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+            for evt in &self.events {
+                write!(f, "{evt}; ")?;
+            }
+            Ok(())
+        }
+    }
+
+    impl LocStateProtPair {
+        /// List all sequences of operations that lead to non-UB configurations,
+        /// without duplicates (not guaranteed to be the shortest sequence of events).
+        fn applicable_opaque_code_without_ub<RetX, RetY>(
+            self,
+            xy_rel: RelPosXY,
+        ) -> impl Iterator<Item = (Self, OpaqueCode<RetX, RetY>)>
+        where
+            RetX: Exhaustive + Clone + 'static,
+            RetY: Exhaustive + Clone + 'static,
+        {
+            // (state, code) left to handle
+            let mut handle = vec![(self, OpaqueCode { events: Vec::new() })];
+            // code that can be applied to `self`, and final state
+            let mut paths = Vec::new();
+            // already explored states reachable from `self`
+            let mut seen = FxHashSet::default();
```

Comment:
Please add a type annotation, it's a bit hard to figure out which kinds of elements this set has.